cpp/src/io/json/json_tree.cu

```diff
@@ -0,0 +1,661 @@
+                                   auto const& node_type,
+                                   auto const& levels,
+                                   auto const& col_id) {
+    std::cout << level << stage_text << "\n";
+    auto cls = std::array<const char*, 6>{"S", "L", "F", "R", "V", "E"};
+    for (auto n = start; n < end; n++)
+      printf("%3d ", nodeII.element(n, stream));
+    printf(" nodeII-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", parent_node_idx.element(n, stream));
+    printf(" parent_node_idx-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", parent_col_id.element(n, stream));
+    printf(" parent_col_id-%ld\n", level);
+    for (auto n = start; n < end; n++) {
+      auto nt = node_type.element(n, stream);
+      printf("%3s ",
+             nt >= NUM_NODE_CLASSES ? std::to_string(nt - NUM_NODE_CLASSES).c_str() : cls[nt]);
+    }
+    printf(" node_type-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", levels.element(n, stream));
+    printf(" levels-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", col_id.element(n, stream));
+    printf(" col_id-%ld\n", level);
+  };
+#define PRINT_LEVEL_DATA(level, text)           \
+  print_level_data(text,                        \
+                   level,                       \
+                   level_boundaries[level - 1], \
+                   level_boundaries[level],     \
+                   scatter_indices,             \
+                   parent_indices,              \
+                   parent_col_id,               \
+                   node_type,                   \
+                   d_tree.node_levels,          \
+                   col_id);
+#else
+#define PRINT_LEVEL_DATA(level, text) ;
+#endif
+
+  // 4. Propagate parent node ids for each level.
+  // For each level,
+  //     a. gather col_id from previous level results. input=col_id, gather_map is parent_indices.
+  //     b. stable sort by {parent_col_id, node_type}
+  //     c. scan sum of unique {parent_col_id, node_type}
+  //     d. scatter the col_id back to stable node_level order (using scatter_indices)
+  rmm::device_uvector<NodeIndexT> col_id(num_nodes, stream);
+  rmm::device_uvector<NodeIndexT> parent_col_id(num_nodes, stream);
+  thrust::uninitialized_fill(rmm::exec_policy(stream),
+                             parent_col_id.begin(),
+                             parent_col_id.end(),
+                             0);  // XXX: is this needed?
+  // fill with 1, useful for scan later
+  thrust::uninitialized_fill(rmm::exec_policy(stream), col_id.begin(), col_id.end(), 1);
+  // TODO: Could initialize to 0 and scatter to level_boundaries
+  thrust::device_pointer_cast(col_id.data())[0] = 0;  // Initialize First node col_id to 0
+  thrust::device_pointer_cast(parent_col_id.data())[0] =
+    -1;  // Initialize First node parent_col_id to -1 sentinel
```

Comment:
Same here to avoid superfluous synchronisation
```suggestion
  col_id.data.set_element_async(0, 0, stream);  // Initialize First node col_id to 0
  parent_col_id.set_element_async(0, no_parent_sentinel, stream);
```

Reply:
needed to replace this with `thrust::fill` for json-lines support.